### PR TITLE
Refine documentation

### PR DESCRIPTION
**Pull request by kellertuer**
This small PR extends the documentation especially of exp and retract. Maybe this is a good PR to also improve other docstrings, so feel free to propose further changes.

### CODE REVIEW

**Comment by mateuszbaran** [comment]
Since we already work on this, maybe it would be nice to also point out the difference between a geodesic and a curve that minimizes distance?

**Comment by kellertuer** [comment]
Good idea, that could best be done in `geodesic` and `shortest_geodesic` then.

**Comment by kellertuer** [comment]
I added more details on the geodesic. Since we are at it, we could also add ApproximateInverseRetraction (and by the way an approximate log would also be neat), and solve JuliaManifold/Manifolds.jl#343 as well by adding the `NLsolveInverseRetraction` here, too, though then I would maybe add it here and put retractions into their separate section in the interface.html?.

**Comment by mateuszbaran** [comment]
Sure, I think a general section about retractions, inverse retractions and projections (with some figures as discussed in https://github.com/JuliaNLSolvers/Optim.jl/issues/920 ) would be helpful.

**Comment by Affie** [comment]
> Sure, I think a general section about retractions, inverse retractions and projections (with some figures as discussed in JuliaNLSolvers/Optim.jl#920 ) would be helpful.

That sounds great. The optim.jl documentation gave me the wrong idea that `retract` and `project_tangent` were opposites.
So it should help users in the future to have retractions and inverse retractions defined together with the projections.

**Comment by kellertuer** [comment]
Well, Optim.jl uses the retraction as it is defined for topological spaces for a retraction on a manifold, where it has a different meaning usually, for that we can‘t do anything.

For the figure I have to take a look what would be a suitable illustration. Currently thinking of maybe even using just the Circle.

**Comment by Affie** [comment]
I drew it in my notebook as Circles, I have about 6 of them :-)

**Comment by kellertuer** [comment]
Hm, I somehow broke something (ExponentialRetraction is not found anymore) though I just copied (carefully) things around. Will take a look later.

**Comment by kellertuer** [comment]
I did a minor bump, since we reorganised files here (and Manifolds.jl has to adapt to that), but it is non-breaking though I also moved two types here (which breaks nothing here and when upgrading manifolds will also break nothing there).

**Comment by kellertuer** [comment]
Just as a remark, the last commit puts the implementation of the retraction with NLsolve back into Manifolds.jl, since it would introduce a dependency on `Reuiqres.jl` here.

**Review by mateuszbaran — COMMENTED** [review]
I haven't reviewed everything yet but here are some small points.

File: Project.toml
Comment (kellertuer) [reply Project.toml:4]: I maybe forgot how to count.

File: src/DecoratorManifold.jl
Comment (kellertuer) [reply src/DecoratorManifold.jl:577]: But then I don‘t know how and where do document it? Currently the type is here and the retract (since it needs NLSolve) is in manifolds.

File: src/DefaultManifold.jl
Comment (kellertuer) [reply src/DefaultManifold.jl:64]: Ok, I just saw its not used and get a VS code Liter warning, but you are right, without a type its better to keep `p`

File: src/DecoratorManifold.jl
Comment (mateuszbaran) [reply src/DecoratorManifold.jl:577]: Why does the type need to be here?

File: src/PowerManifold.jl
Comment (kellertuer) [reply src/PowerManifold.jl:659]: Yes,, sorry, I moved retract over so I had to add this, moved retract back already for NLsolveInverseRetraction

File: src/DecoratorManifold.jl
Comment (kellertuer) [reply src/DecoratorManifold.jl:577]: That‘s what we thought a good idea in JuliaManifolds/Manifolds.jl#343. If we do not move it here I have no clue where to put the documentation, actually.

File: src/ManifoldsBase.jl
Comment (kellertuer) [reply src/ManifoldsBase.jl:217]: Sure.

File: src/exp_log_geo.jl
Comment (kellertuer) [reply src/exp_log_geo.jl:67]: Sure. I also `@ref`ed `retract` in `exp`, too, now..

File: src/DecoratorManifold.jl
Comment (mateuszbaran) [reply src/DecoratorManifold.jl:577]: Right, I forgot about it.

Comment (kellertuer) [reply src/DecoratorManifold.jl:577]: I know this is not the optimal solution – providing a type here and implementing the retraction only in Manifolds, but I have no better idea currently.

File: src/exp_log_geo.jl
Comment (mateuszbaran) [reply src/exp_log_geo.jl]: @kellertuer What do you think about my comment above about the word "shortest"

Comment (kellertuer) [reply src/exp_log_geo.jl]: Ah, yes, we cam commit the shortest, sure.

File: src/maintypes.jl
Comment (mateuszbaran) [inline src/maintypes.jl:16]: Maybe you could `@ref` the section in docs about implementing a new manifold?

Comment (kellertuer) [reply src/maintypes.jl:16]: Interesting idea. My only point, where I am unsure, is, that if you call this in REPL, you would just see "How to implement a manifold"?

File: src/projections.jl
Comment (mateuszbaran) [inline src/projections.jl:47]: Is projecting a vector using this and projecting a point from the tangent space manifold using like in https://github.com/JuliaNLSolvers/Optim.jl/issues/920#issuecomment-841662515 the same thing or not? Intuitively projecting to the tangent space may not entail representation change.

File: src/maintypes.jl
Comment (mateuszbaran) [reply src/maintypes.jl:16]: No, that's not what I had in mind. I think that `@ref` could be added at the bottom of the docs, and not replace the text that's here.

File: src/projections.jl
Comment (kellertuer) [reply src/projections.jl:47]: I don‘t get your second point; If you implement `TangentSpaceAtPoint` to use the same representation as the tangent space at `p` in the usual sense, then the projection of the first should also change representation...?

File: src/maintypes.jl
Comment (kellertuer) [reply src/maintypes.jl:16]: Yes, that‘s what I understood :) Still in the REPL this last (additional) line would not help that much since its not a link there?

File: src/projections.jl
Comment (mateuszbaran) [reply src/projections.jl:47]: Intuitively, to be consistent, the rules for projection onto `TangentSpaceAtPoint` should follow the rules for projection onto a manifold.

Before I continue, do you think we should have for points `p that are `is_point(M::AbstractEmbedded, p)==true`  the equality `project(M::AbstractEmbedded, p) == p`? It's very easy to assume that's true but representation change would violate this. In any case, I think this should be documented.

File: src/maintypes.jl
Comment (mateuszbaran) [reply src/maintypes.jl:16]: It would be helpful when browsing the online docs. Additionally we can have a URL in the docstring that would be helpful in REPL.

File: src/projections.jl
Comment (mateuszbaran) [reply src/projections.jl:47]: Also, at the moment `embed` for `Rotations` doesn't revert the transformation but it seems like it should.

Comment (kellertuer) [reply src/projections.jl:47]: Yes, but that's more like something over on Manifolds ;)

File: src/maintypes.jl
Comment (kellertuer) [reply src/maintypes.jl:16]: Sure. I am not 100% a fan of that, since the URL would be static (e.g. also link to the stable tutorial when looking at the dev variant, but we could do that.

Comment (kellertuer) [reply src/maintypes.jl:16]: We could also add it to the `interface.md` just before the docs block containing this type?

File: src/projections.jl
Comment (mateuszbaran) [reply src/projections.jl:47]: Fixing that problem is something for Manifolds.jl, right, and I think we still need to discuss some points about projections and embeddings but maybe let's leave it to a different PR.

File: src/maintypes.jl
Comment (mateuszbaran) [reply src/maintypes.jl:16]: Add what to `interface.md`? I'd just like to have something that would point people who read the docstring of `Manifold` type to `interface.md`.

Comment (kellertuer) [reply src/maintypes.jl:16]: Ah, I though something that links to the tutorial, that could be done in interface. Hm, then I will think of something.

File: src/retractions.jl
Comment (mateuszbaran) [inline src/retractions.jl:175]:
```suggestion
For available inverse retractions on certain manifolds see the documentation on the
```

File: src/retractions.jl
Comment (mateuszbaran) [inline src/retractions.jl:167]:
```suggestion
    inverse_retract(M::Manifold, p, q)
    inverse_retract(M::Manifold, p, q, method::AbstractInverseRetractionMethod
```

File: src/retractions.jl
Comment (mateuszbaran) [inline src/retractions.jl:200]: I think you forgot the codomain of retraction here.

File: src/retractions.jl
Comment (mateuszbaran) [inline src/retractions.jl:203]: I think you mix here `\operatorname{retr}` with `R`, and you could add a note about what the `D` is.

**Review by mateuszbaran — COMMENTED** [review]
A few more comments :slightly_smiling_face: .

File: src/retractions.jl
Comment (kellertuer) [reply src/retractions.jl:203]: Thanks I added both.

File: src/retractions.jl
Comment (kellertuer) [reply src/retractions.jl:200]: Ah, yes, sure, I added the codomain.

**Comment by kellertuer** [comment]
Oh, this was the first time I actually had to carefully do a merge conflict resolution, since we edited the same areas. I hope though that I did not miss anything and we might face something similar in Manifolds.jl then, though there it might be less severe, since I mainly edited docs.

**Comment by mateuszbaran** [comment]
> Oh, this was the first time I actually had to carefully do a merge conflict resolution, since we edited the same areas. I hope though that I did not miss anything and we might face something similar in Manifolds.jl then, though there it might be less severe, since I mainly edited docs.

I'll review the changes again :slightly_smiling_face: .

> **edit:** A I might have to rethink my section on main types, which covered Manifold, MPoint, TVector and CoTVector, but maybe the last two are now moved to a vector space section anyways?

Yes, the `TVector`/`CoTVector` thing was changed a bit (and we now have `AbstractFibreVector`). It may make more sense to talk about them in a vector space section.

**Comment by kellertuer** [comment]
Ah, sure, that makes more sense. I will wait for your manifolds branch to be merged and then also there adapt the docs in my rewrite.

**Comment by dehann** [comment]
Thanks for all the effort, these documentation refinements really help a lot!

**Comment by kellertuer** [comment]
I think this one is also good to be merged. I have locally checke both this and the manifolds branch with the new (0.11/0.5) versions and locally all looks good.

**Comment by mateuszbaran** [comment]
The failure on nightly is probably caused by https://github.com/JuliaLang/julia/pull/33697 .

File: src/exp_log_geo.jl
Comment (mateuszbaran) [inline src/exp_log_geo.jl:11]:
```suggestion
where ``γ_{p,X}`` is the unique geodesic starting in ``γ(0)=p`` such that ``\dot γ(0) = X``.
```

File: src/exp_log_geo.jl
Comment (mateuszbaran) [inline src/exp_log_geo.jl:9]:
```suggestion
\exp_p X = γ_{p,X}(1),
```
I think it will be more readable in REPL with a space there (I know this makes no difference to LaTeX rendering).

**Comment by kellertuer** [comment]
That is a _very_ neat notation they came up with, I just do net yet see why that breaks our `show`s ;) but yes, that seems related.

File: src/vector_transport.jl
Comment (mateuszbaran) [inline src/vector_transport.jl:84]: Should we add a symbol for covariant derivative to `notation.md`?

**Review by mateuszbaran — APPROVED** [review]
These changes look nice :+1:

File: src/exp_log_geo.jl
Comment (kellertuer) [reply src/exp_log_geo.jl:9]: Ah, sure, I did not think about REPL here.

File: src/vector_transport.jl
Comment (kellertuer) [reply src/vector_transport.jl:84]: That sounds reasonable, I will add that.

Comment (kellertuer) [reply src/vector_transport.jl]: I added it and running docs locally just now. Note that this is of course added over on Manifolds.jl.

Comment (mateuszbaran) [reply src/vector_transport.jl]: Great, thanks!

**Pull request merged**
